traits/trait_list_object.py

```diff
@@ -53,460 +88,755 @@ def __init__(self, index=0, removed=None, added=None):
-                            return
-                    except Exception:
-                        # Treat incomparable values as equal:
-                        pass
-                self._send_trait_items_event(
-                    self.name_items, TraitListEvent(index, removed, values)
-                )
+    def __getstate__(self):
+        """ Get the state of the object for serialization.
 
-        except TraitError as excp:
-            excp.set_prefix("Each element of the")
-            raise excp
+        Notifiers are transient and should not be serialized.
+        """
+        result = self.__dict__.copy()
+        result.pop("notifiers", None)
+        return result
 
-    def __delitem__(self, key):
-        trait = getattr(self, "trait", None)
-        if trait is None:
-            return list.__delitem__(self, key)
+    def __setstate__(self, state):
+        """ Restore the state of the object after serialization.
 
-        try:
-            removed = self[key]
-        except Exception:
-            removed = []
+        Notifiers are transient and are restored to the empty list.
+        """
+        state['notifiers'] = []
+        self.__dict__.update(state)
 
-        if isinstance(key, slice):
-            slice_len = len(removed)
-            delta = slice_len
-            step = 1 if key.step is None else key.step
-            if step == 1:
-                # FIXME: See corresponding comment in __setitem__() for
-                # explanation.
-                index = 0 if key.start is None else key.start
-            else:
-                index = key
-        else:
-            delta = 1
-            index = len(self) + key + 1 if key < 0 else key
-            removed = [removed]
+    def __setitem__(self, index, value):
+        """ Set a value at index, implements self[index] = value
+
+        Parameters
+        ----------
+        index : int
+            Index at which the value will be set.
+        value : Any
+            The value to set at the index.
+
+        Notes
+        -----
+        Notification:
+            index : int
+                The index of the value.
+            added : Any
+                The value that is set.
+            removed : Any
+                The removed value or Undefined if no value was previously
+                present at the index.
 
-        if not (trait.minlen <= (len(self) - delta)):
-            self.len_error(len(self) - delta)
-            return
+        """
+        removed = self._get_removed(index)
 
-        list.__delitem__(self, key)
+        if isinstance(index, slice):
+            if len(removed) != len(value) and index.step not in {1, None}:
+                raise ValueError
 
-        if self.name_items is not None and removed:
-            self._send_trait_items_event(
-                self.name_items, TraitListEvent(index, removed)
-            )
+        added = self.validate(index, removed, value)
```

Comment:
See comment above about always sending a list for `removed` and `added`.  This `value` should be rewrapped as `[value]` if `index` is an integer.

In case `value` is an iterable, it should be converted to a list before being reused in `validate` and `notify`.

Similarly, other methods are affected too, e.g. `__delitem__`.